### PR TITLE
feat: X.509 service

### PR DESCRIPTION
**Pull request by berendsliedrecht**
- Create X.509 certificates based on a key
- Validate an X.509 Certificate chain
- Moved hashing libs from stablelib -> noble
- Added sha-1 hashing method
- Added a webcrypto implementation that uses the credo wallet for operations
  - Support for P256, P384, K256 and Ed25519

~~In draft for now as the webcrypto implementation won't work well in production. The issue is is that we have to map webcrypto key algorithm (signing, verifying, import, export, etc.) to credo (and possibly also reverse). The reverse is very difficult, unless we assume askar and that we always use SHA-256 when signing with P-256.~~

I see that cheqd has a dependency on `@stablelib/ed25519` and core as well. Might be good to see whether we can move that when we have a full webcrypto implementation or just use `@nobile/curves` for that.

### CODE REVIEW

File: packages/core/src/modules/sd-jwt-vc/SdJwtVcService.ts
Comment (TimoGlastra) [inline packages/core/src/modules/sd-jwt-vc/SdJwtVcService.ts:445]: should sanDnsNames and sanUriNames return an empty array if there's no entries? That would make this code cleaner. Or is there different in meaning from retruning an empty array vs undefined?

File: packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts:414]: still needs to be supported in sphereons libs?

File: packages/openid4vc/src/shared/models/OpenId4VcJwtIssuer.ts
Comment (TimoGlastra) [inline packages/openid4vc/src/shared/models/OpenId4VcJwtIssuer.ts:35]:
```suggestion
export type OpenId4VcJwtIssuer = OpenId4VcJwtIssuerDid | OpenId4VcJwtIssuerX5c
```

File: packages/openid4vc/src/shared/models/OpenId4VcJwtIssuer.ts
Comment (TimoGlastra) [inline packages/openid4vc/src/shared/models/OpenId4VcJwtIssuer.ts:25]: > Must be provided when creating x509 certificate protected authorization requests.

We're already in x5c method here. So should this be required then? In which case is it optional? For the issuance side it is not optional?

File: packages/openid4vc/src/shared/utils.ts
Comment (TimoGlastra) [inline packages/openid4vc/src/shared/utils.ts:66]: array can be empty. We should check if alg is defined

File: packages/openid4vc/src/shared/utils.ts
Comment (TimoGlastra) [inline packages/openid4vc/src/shared/utils.ts:89]: Hmm we shouldn't mere it like this. Do we want to wait until all the updates to Sphereon libs are merged/released? Or can we temp disable x5c until this is fixed so we can already merge this work? Woudln't be a bummer to keep this PR open for too long

File: packages/core/src/modules/sd-jwt-vc/SdJwtVcService.ts
Comment (TimoGlastra) [inline packages/core/src/modules/sd-jwt-vc/SdJwtVcService.ts:406]: We should check if the index 0 exists (could be undefined). Maybe we can add a `firstSupportedSignatureAlgorithm` that throws if not found? That way we don't have to do the check everywhere

**Review by TimoGlastra — CHANGES_REQUESTED** [review]
Good stuff.

I'm curious, are there any breaking changes in this PR? Would be nice if it doesn't, but otherwise we have to think about versioning (and maybe merge some parts into a separate branch)

**Comment by TimoGlastra** [comment]
Also -- DCO is failing

File: packages/core/src/crypto/webcrypto/CryptoKey.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/webcrypto/CryptoKey.ts:6]: I wanted to avoid the Credo prefix, but not really for any major specific reason. Can add it or keep it at `WebCryptoKey` as well. If we prefix it here with `Credo` I would like to do it for all the classes.

File: packages/core/src/crypto/webcrypto/WalletWebCrypto.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/webcrypto/WalletWebCrypto.ts:40]: Yeah I think that should be possible. We can also just add a new `getRandomValues` method to the wallet and use that. `generateNonce` can stay async.

File: packages/core/src/crypto/__tests__/JwsService.test.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/__tests__/JwsService.test.ts:54]: I don't know why this test is deleted. I think this might have been @auer-martin?

File: packages/core/src/crypto/webcrypto/utils/keyAlgorithmConversion.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/webcrypto/utils/keyAlgorithmConversion.ts:12]: Yeah that was my idea as well. Its an "easy" mapping, but a bit time consuming so I wanted to do it when everything was working with ES256.

File: packages/core/src/crypto/x509/X509Certificate.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/x509/X509Certificate.ts:21]: Yes!

File: packages/core/src/crypto/x509/X509Certificate.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/x509/X509Certificate.ts:49]: Yeah it will break, but this should actually use the conversion utils, good catch!

File: packages/core/src/crypto/x509/X509Certificate.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/x509/X509Certificate.ts:55]: Yes, will also be caught by proper conversion.

File: packages/core/src/crypto/x509/X509Certificate.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/x509/X509Certificate.ts:117]: The underlying `Key` instance for the public key is the exact same as the private key as they are connected in a lower level. I think if we merge the `Key` and the `CryptoKey` more in the future it will be better, I get that this looks a bit weird, but it is.

File: packages/core/src/crypto/x509/X509Service.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/x509/X509Service.ts:16]: If we pass the `agentContext` into every method it has, yes this is easily doable!

Comment (berendsliedrecht) [reply packages/core/src/crypto/x509/X509Service.ts:16]: Also, this is in another case the wrong place. If we want to reuse the webcrypto in other services we should register it at a higher level. But I will pass the `agentContext` to each service call and inject it there into the method.

File: packages/core/src/crypto/x509/X509Service.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/x509/X509Service.ts:19]: An array of like:

```json
['-----BEGIN ....', '-----BEGIN ....', ...]
````

Thats possible, but I might miss the exact use case for that.

File: packages/core/src/crypto/x509/__tests__/x509Service.test.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/x509/__tests__/x509Service.test.ts:43]: Yes!

File: packages/core/src/modules/dids/methods/peer/peerDidNumAlgo1.ts
Comment (berendsliedrecht) [reply packages/core/src/modules/dids/methods/peer/peerDidNumAlgo1.ts:7]: Yes, if the multibase encoder is exported (which I am fairly sure it is) this will break on types. 

Do we want to keep it as `sha-256` or support `SHA-256` | `sha-256` | `sha256`?

File: packages/core/src/modules/sd-jwt-vc/SdJwtVcOptions.ts
Comment (berendsliedrecht) [reply packages/core/src/modules/sd-jwt-vc/SdJwtVcOptions.ts:46]: @auer-martin comment for you :).

File: packages/core/src/modules/sd-jwt-vc/SdJwtVcService.ts
Comment (berendsliedrecht) [reply packages/core/src/modules/sd-jwt-vc/SdJwtVcService.ts:445]: Technically there is a difference, i.e. undefined = no extension for SAN, [] = extension for SAN was there, but did not contain any values.

I don't think we want to make a distinction between these cases, so I will default to an empty array if nothing is found.

File: packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts
Comment (berendsliedrecht) [reply packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts:414]: @auer-martin

File: packages/openid4vc/src/shared/models/OpenId4VcJwtIssuer.ts
Comment (berendsliedrecht) [reply packages/openid4vc/src/shared/models/OpenId4VcJwtIssuer.ts:35]: @auer-martin

File: packages/openid4vc/src/shared/models/OpenId4VcJwtIssuer.ts
Comment (berendsliedrecht) [reply packages/openid4vc/src/shared/models/OpenId4VcJwtIssuer.ts:25]: @auer-martin  :)

File: packages/openid4vc/src/shared/utils.ts
Comment (berendsliedrecht) [reply packages/openid4vc/src/shared/utils.ts:66]: @auer-martin I think you get it, haha. All the comments outside of the `crypto` folder are for you :D.

File: packages/core/src/crypto/x509/X509Service.ts
Comment (TimoGlastra) [reply packages/core/src/crypto/x509/X509Service.ts:19]: No a string like this:

```
BEGIN CERTIFICATE

END CERTIFICATE
BEGIN CERTIFICATE
END CERTIFICATE
```

Thats how you usually store certificates in a file

Comment (TimoGlastra) [reply packages/core/src/crypto/x509/X509Service.ts:19]: I assume certificates for issuance will be created outside of credo, and imported based on the above format

File: packages/core/src/modules/dids/methods/peer/peerDidNumAlgo1.ts
Comment (TimoGlastra) [reply packages/core/src/modules/dids/methods/peer/peerDidNumAlgo1.ts:7]: Let's for now do backwards compat (so also lowercase sha-256). For sha-1 we only have to support the upper case variant. In 0.6 we can remove the lower case variant

File: packages/core/src/crypto/x509/X509Service.ts
Comment (TimoGlastra) [reply packages/core/src/crypto/x509/X509Service.ts:16]: Okay good. and it won't overwrite it for the whole x509 package? As the current method sets the crypto provider globally right?

**Comment by TimoGlastra** [comment]
Also would be good if we can create some changesets for this (i can add it). What are the changes we want to document:
- add support for x5c issuer method for SD-JWT VCs to issue a credential linked to a x5c certificate chain
- add support for `x509_san_dns` and `x509_san_uri` client_id_schemes to SIOPv2/OpenID4VP for signing authorization request JWT signer methods 
- add support for `x5c` in JWSService
- add X509Service for parsing and verifying X509 Certificate Chains and adds an util to create self-signed certificates

File: packages/core/src/modules/dids/methods/peer/peerDidNumAlgo1.ts
Comment (berendsliedrecht) [reply packages/core/src/modules/dids/methods/peer/peerDidNumAlgo1.ts:7]: Kept it as `sha-256` and when checking it, I do `hashName.toUpperCase() === 'SHA-256'`

**Comment by berendsliedrecht** [comment]
> Also would be good if we can create some changesets for this (i can add it). What are the changes we want to document:
> 
> * add support for x5c issuer method for SD-JWT VCs to issue a credential linked to a x5c certificate chain
> * add support for `x509_san_dns` and `x509_san_uri` client_id_schemes to SIOPv2/OpenID4VP for signing authorization request JWT signer methods
> * add support for `x5c` in JWSService
> * add X509Service for parsing and verifying X509 Certificate Chains and adds an util to create self-signed certificates
> 
> Anything missing?

From a public perspective? I think that covers it. However, I would like to merge the `crypto` side and let @auer-martin continue in a new PR. I don't want this to stay open for too long and deal with merge conflicts haha. I can revert all the changes that are not in the crypto package.

File: packages/core/src/crypto/x509/X509Service.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/x509/X509Service.ts:16]: Right now, in the new implementation I set it and clear it. AFAIK its a local global, not `global.crypto` directly.

File: packages/core/src/crypto/x509/X509Service.ts
Comment (berendsliedrecht) [reply packages/core/src/crypto/x509/X509Service.ts:19]: Ah Yeah that could be useful!

**Comment by berendsliedrecht** [comment]
@TimoGlastra I only included the crypto changes in this PR for now. Only TODO is the conversion of crypto algorithm, spki algorithm and credo algorithm.

File: packages/core/src/crypto/__tests__/JwsService.test.ts
Comment (auer-martin) [reply packages/core/src/crypto/__tests__/JwsService.test.ts]: Sorry, then it probably was a mistake on my end.

File: packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-verifier/OpenId4VcSiopVerifierService.ts]: yes needs to be implemented

File: packages/core/src/crypto/x509/X509Service.ts
Comment (TimoGlastra) [reply packages/core/src/crypto/x509/X509Service.ts]: Yes but that could still cause conflicts with async methods doing multiple operations right?

Because if method X is called twice concurrently. After doing one operation, method x needs to wait for an async operation, and then the other x method is ran. Now the x509 crypto will be overriden. And when the first call to X resumes it will have another crypto instance, possibly bound to another agent context?

Or are all calls to x509 library sync and do we set it before each call to x509 library?

File: packages/askar/src/wallet/AskarBaseWallet.ts
Comment (TimoGlastra) [inline packages/askar/src/wallet/AskarBaseWallet.ts:434]: I think it's more Credo-like to pass in a value length and get back an UInt8array instead of passing in a uint8array with length

File: packages/core/src/crypto/x509/X509Certificate.ts
Comment (TimoGlastra) [reply packages/core/src/crypto/x509/X509Certificate.ts:50]: Still todo? or now fixes with the conversion utils?

File: packages/core/src/crypto/x509/X509Certificate.ts
Comment (TimoGlastra) [reply packages/core/src/crypto/x509/X509Certificate.ts:56]: this as well?

File: packages/core/src/crypto/x509/X509Service.ts
Comment (TimoGlastra) [inline packages/core/src/crypto/x509/X509Service.ts:40]: if user passes an empty `cerificateChain`, but also a string to `certificate` this will not throw, even if the `certificateChain` is empty

File: packages/core/src/crypto/x509/X509Service.ts
Comment (TimoGlastra) [inline packages/core/src/crypto/x509/X509Service.ts:64]: I think this is still a problem with the global crypto provider that is configured... As we can't be sure another process doesn't override the x509 crypto provider as we call async/await

**Review by TimoGlastra — APPROVED** [review]

**Pull request merged**
